### DISCUSSION TITLE
`page.mdx` in the `app` directory.

### DISCUSSION THREAD

**Discussion by Lakuna** [vercel/next.js#42882] [Ideas]
### Describe the feature you'd like to request

In Next.js 13's new `app` directory, when using `@next/mdx`, it is currently necessary to use a `page.jsx` file to display content in a `.mdx` file:

`app/content.mdx`

```mdx
# Example

Lorem ipsum dolor sit amet.
```

`app/page.jsx`

```jsx
import Content from "./content.mdx";

export default function Page() {
  return <Content />;
}
```

The `page.jsx` could be removed from cases like this by allowing `page.mdx` to be used as the page file.

`app/page.mdx`

```mdx
# Example

Lorem ipsum dolor sit amet.
```

### Describe the solution you'd like

The `page.jsx` could be removed from cases like this by allowing `page.mdx` to be used as the page file.

`app/page.mdx`

```mdx
# Example

Lorem ipsum dolor sit amet.
```

### Describe alternatives you've considered

The current implementation works fine, but requires the use of `page.jsx` files that don't add any content to the page in some situations. I believe that the implementation described above would fit well with the new `app` directory structure introduced in Next.js 13, and is more intuitive than the current implementation.

**Comment by bfourgeaud** [comment]
Hello

I have tried to use `@next/mdx` while converting a project to next13 but I havn't been ble to make it work, even with your solution.

`next.config.js`
```js
const withMDX = require('@next/mdx')({
  extension: /\.mdx?$/,
})

module.exports = withMDX({
  pageExtensions: ['ts', 'tsx', 'js', 'jsx', 'md', 'mdx'],
  reactStrictMode: true,
  swcMinify: false, // Required to fix: https://nextjs.org/docs/messages/failed-loading-swc
  experimental:{
    appDir: true
  }
})
```

`app/page.tsx`
```tsx
import Content from './content.mdx'

export default function Reseau() {
  return <Content />
}
```

`app/content.mdx`
```mdx
# Example

Lorem ipsum dolor sit amet.
```

## Error code displayed when visiting page.tsx
`TypeError: react__WEBPACK_IMPORTED_MODULE_0__.createContext is not a function`


Do you have an idea what I am doing wrong ?
Thanks,

**Reply by Lakuna** [reply]
Yes; see #42153.

**Reply by bfourgeaud** [reply]
Perfect it worked like a charm ;)
Now just have to wait to have the possibility to directly add `page.mdx` files ;)

**Reply by XDean** [reply]
I find a workaround to add `use client` to mdx. See [sample project](https://github.com/XDean/next-13-mdx)

**Comment by tubbo** [comment]
This would be fantastic. It could potentially use the `pageExtensions` config (but only be applicable for page files).

**Comment by y-nk** [comment]
There's the possibility (although ugly) to manually load markdown on the server without the use of webpack (since we operate server-side).

1. We create a dummy page which will load our md content. The component is a simple proxy calling `await getPost(params.slug)` and rendering the returned component.

```tsx:/app/[slug]/page.tsx
import { getSlugs, getPost } from "./posts";

export const dynamicParams = false

export async function generateStaticParams() {
  const slugs = await getSlugs()
  return slugs.map(slug => ({ slug }));
}

export default async function Post({ params }: { params: any }) {
  const { Content, metadata } = await getPost(params.slug) // ← here

  return (
    <main>
      <Content />
    </main>
  )
}
```

2. We create the utils necessary to load the md.

```ts:/app/[slug]/posts.tsx
import { readdir, readFile } from "fs/promises";
import { dirname, join } from "path";

import type { FC } from 'react'

import { unified } from "unified";
import remarkParse from "remark-parse";
import remarkRehype from "remark-rehype";
import rehypeReact from "rehype-react";
import rehypeStringify from "rehype-stringify";

type Post = {
  Content: FC,
  metadata: {
    [x: string]: string
  }
}

const DIRNAME = dirname(
  new URL(import.meta.url).pathname
)

const IGNORE_FILES = ['page.tsx', 'posts.tsx']

async function loadModule(slug: string) {
  const {
    default: Content,
    metadata = {},
  } = await import(`./${slug}`)

  return {
    Content: () => (
      <article><Content /></article>
    ),
    metadata,
  } as Post
}

async function loadMarkdown(slug: string, mdx = false) {
  const indexPath = join(DIRNAME, slug, `index.md${mdx ? 'x' :''}` )
  const fileContent = await readFile(indexPath, 'utf-8')

  const { value } = await unified()
    .use(remarkParse)
    .use(remarkRehype, { allowDangerousHtml: true })
    .use(rehypeStringify, { allowDangerousHtml: true })
    .process(fileContent)

  return {
    Content: () => (
      <article dangerouslySetInnerHTML={{ __html: `${value}` }} />
    ),
    metadata: {},
  } as Post
}

export async function getSlugs() {
  const files = await readdir(DIRNAME)
  return files.filter(file => !IGNORE_FILES.includes(file))
}


export async function getPost(slug: string) {
  const files = await readdir(join(DIRNAME, slug))
  const index = files.find(file => file.startsWith('index.')) ?? ''

  if (/\.tsx?$/i.test(index)) {
    return loadModule(slug)
  } else if (/.mdx?$/i.test(index)) {
    return loadMarkdown(slug, index.endsWith('x'))
  } else {
    throw new Error(`unknown post: ${slug}`)
  }
}
```

The `getPost` function will test and see if it's necessary to load script or markdown content ; for our example it'll call `loadMarkdown` which will read the file manually, pass the content through a unified pipeline then use `dangerouslySetInnerHTML` to inject the md content in the page.

I've added `allowDangerousHtml` in unified plugins to allow `<style>` and `<script>` client-side. This works ok for markdown but since we're not using webpack, the import in `loadModule` does not pull other files such as css modules - i would assume it'll be the same with mdx imports.

**Reply by nishuzumi** [reply]
This has a fatal problem, when you modify the content, it does not trigger a hot update

**Reply by y-nk** [reply]
@nishuzumi indeed but there's nothing we can do about this since the bundler isn't exposed to us :( for HMR to work, we'd need native support of mdx in app/

**Comment by pomber** [comment]
`app/page.mdx` works now (v13.1.6)

see [Markdown and MDX beta docs](https://beta.nextjs.org/docs/guides/mdx)

**Reply by morgs32** [reply]
Are you sure that remark and rehype plugins work in this new paradigm?

**Reply by morgs32** [reply]
The answer to that:

Looks like mdx support with "app dir" is provided by [mdxjs-rs](https://github.com/wooorm/mdxjs-rs) and:
> This project does not yet support plugins. To benefit from the unified (remark and rehype) ecosystem, use [@mdx-js/mdx](https://mdxjs.com/packages/mdx/).

**Reply by timneutkens** [reply]
It can work with the JS version too but we'd have to upgrade mdx-js/loader https://github.com/vercel/next.js/blob/canary/packages/next-mdx/package.json#L14 which is a breaking change

**Reply by morgs32** [reply]
I'm curious about the implications of that, could you say more? Would a major update to next-mdx spill over into nextjs? Or is it just we should look for a feasible solution in some new features (in which case, could you say more about that technical approach?)

**Reply by villasv** [reply]
And just to clarify, "works now" is just "works as the 'current behavior' described in the issue, but the feature being requested is still not available", right? Because the linked docs and example project still wrap the MDX in a component instead of rendering a page directly.

**Reply by graydonpleasants** [reply]
In my testing this does not work still, though my test `.mdx` pages are in a subfolder within `/app`

**Reply by beeinger** [reply]
Btw in [these docs](https://nextjs.org/docs/app/building-your-application/configuring/mdx#frontmatter) it should be:
```ts
export const metadata = {
  author: 'Rich Haines',
};
 
# My MDX page
```

not `meta`

**Comment by beeinger** [comment]
Had to play around a lot to get a nice MDX blog with all features I wanted, if anyone is still struggling there is a lot of code you can steal here - https://github.com/beeinger/next-mdx-blog

**Reply by oeyoews** [reply]
Looking at a lot of MDX repositories, I feel that the ecology of MDX is very chaotic, and I will encounter all kinds of strange problems that need to be explored and solved by myself, and I am a little tired now
